src/aof.c

```diff
@@ -40,191 +40,625 @@
+        bg_unlink(ai->file_name);
+        listDelNode(server.aof_manifest->history_aof_list, ln);
+    }
+    server.aof_manifest->dirty = 1;
+    return persistAofManifest(server.aof_manifest);
+}
+
+/* Called after `loadDataFromDisk` when redis start. If `server.aof_state` is 
+ * AOF_ON, It will do two things:
+ * 1. Open the last opened INCR type AOF for writing, If not, create a new one
+ * 2. synchronously update the manifest file to the disk
+ * 
+ * if any of the above two steps fails, the redis process will exit.
+ */
+void aofOpenIfNeededOnServerStart(void) {
+    if (server.aof_state != AOF_ON) {
+        return;
+    }
+
+    serverAssert(server.aof_manifest != NULL);
+    serverAssert(server.aof_fd == -1);
+
+    const char *aof_name = getLastIncrAofName(server.aof_manifest);
+    server.aof_fd = open(aof_name,O_WRONLY|O_APPEND|O_CREAT,0644);
+    if (server.aof_fd == -1) {
+        serverLog(LL_WARNING,"Can't open the append-only file %s: %s",aof_name,strerror(errno));
+        exit(1);
+    }
+
+    int ret = persistAofManifest(server.aof_manifest);
+    if (ret != C_OK) {
+        exit(1);
+    }
+}
+
+/* Called when start to execute AOFRW. If `server.aof_state` is AOF_ON, It
+ * will do two things:
+ * 1. open a new INCR type AOF for writing
+ * 2. synchronously update the manifest file to the disk
+ * 
+ * If any of the above two steps fails, the redis process will exit.
+ * */
+void openNewIncrAofForAppend(void) {
+    serverAssert(server.aof_manifest != NULL);
+
+    sds new_aof_name = getNewIncrAofName(server.aof_manifest);
+
+    /* Close old aof_fd if needed. */
+    if (server.aof_fd != -1) bioCreateCloseJob(server.aof_fd);
+
+    server.aof_fd = open(new_aof_name, O_WRONLY|O_TRUNC|O_CREAT, 0644);
+    if (server.aof_fd == -1) {
+        serverLog(LL_WARNING,"Can't open the append-only file %s: %s",
+            new_aof_name, strerror(errno));
+
+        exit(1);
+    }
+
+    /* Reset the aof_last_incr_size. */
+    server.aof_last_incr_size = 0;
+    int ret = persistAofManifest(server.aof_manifest);
+    if (ret == C_ERR) {
+        exit(1);
     }
-    return count;
 }
 
+/* Whether to limit the execution of Background AOF rewrite.
+ * 
+ * At present, if AOFRW fails, redis will automatically retry. If it continues 
+ * to fail, we may get a lot of very small INCR files. so we need an AOFRW 
+ * limiting measure.
+ * 
+ * We can't directly use `server.aof_current_size` and `server.aof_last_incr_size`, 
+ * because there may be no new writes after AOFRW fails.
+ * 
+ * So, we use time delay to achieve our goal. When AOFRW fails, we delay the execution 
+ * of the next AOFRW by 1 minute. If the next AOFRW also fails, it will be delayed by 2 
+ * minutes. In turn, the maximum delay is 60 minutes (1 hour).
+ * */
+#define AOF_REWRITE_LIMITE_THRESHOLD  16  
+#define AOF_REWRITE_LIMITE_NAX_DELAY  60 /* 1 hour */
+int aofRewriteLimited(void) {
+    int limit = 0;
```

Comment:
@oranagra I implemented this function according to the previous description, because I think we should need it now, because in actual applications AOFRW is often happens (such as  the child OOM), and redis  automatically retry AOFRW is so fast ( default 10HZ), so it is easy to create a bunch of small INCR AOF files.

I am just solving it by increasing the delay linearly (of course there may be other non-linear better algorithms), I think it can solve our problem, if AOFRW has failed 16 times (maybe we need to reduce this value ), then the subsequent AOFRW will become slower and slower.

Reply:
the simple mechanism seems ok, or we can maybe improve (i don't mind too much).
maybe just make it clear in the top comment what's the return value of this function (true to prevent rewrite)

maybe a slightly better one would be to start with a 1 minute delay and then double the delay time on each failure up to a limit of one day?
WDYT?

Reply:
> 16 times (maybe we need to reduce this value

16 seem too high to me too, maybe 8 or even 4 would work?
just remember it can't be lower since we don't count failures, we're counting files, so the base is 3

Reply:
`AOF_REWRITE_LIMITE_THRESHOLD` has been changed to 3, which means that AOFRW may have failed 2 or 3 times continuously. 

`AOF_REWRITE_LIMITE_NAX_DELAY` has been modified to 1440 minutes (1 day). Although I think it may be a bit long, but it is fine.

Reply:
maybe you're right (that a day is too long).
if it fails consistently, then i don't care to delay it for one day, the only problem i see with it is that if the admin fixes the problem (adds more disk / ram space), it'll want to trigger a retry manually ASAP rather than wait for the next interval, and waiting one day may be too long.
so either we add some mechanism to override that (like `BGREWRITEAOF FORCE`), or reduce back to one hour, which may be an acceptable time to wait.
WDYT?

either way, let's be sure to document this in the top comment for the purpose of release notes and other reviewers (which will only look at the description, not the code)

Reply:
I think 1 hour and 1 day are fine, because we can still use the 'bgrewriteaof' command to execute AOFRW immediately during the limit period.

So we reduce back to one hour, just like: 1, 2, 4, 8, 16, 32, 60